OpenRA.Mods.Common/Traits/Player/ClassicParallelProductionQueue.cs

```diff
@@ -0,0 +1,225 @@
+		public readonly int[] ParallelPenalty = { 100, 116, 133, 150, 166, 183, 200, 216, 233, 250 };
+
+		public override object Create(ActorInitializer init) { return new ClassicParallelProductionQueue(init, this); }
+	}
+
+	public class ClassicParallelProductionQueue : ProductionQueue
+	{
+		static readonly ActorInfo[] NoItems = { };
+
+		readonly Actor self;
+		readonly ClassicParallelProductionQueueInfo info;
+
+		int penalty;
+
+		public ClassicParallelProductionQueue(ActorInitializer init, ClassicParallelProductionQueueInfo info)
+			: base(init, init.Self, info)
+		{
+			self = init.Self;
+			this.info = info;
+		}
+
+		protected override void Tick(Actor self)
+		{
+			// PERF: Avoid LINQ.
+			Enabled = false;
+			var isActive = false;
+			foreach (var x in self.World.ActorsWithTrait<Production>())
+			{
+				if (x.Trait.IsTraitDisabled)
+					continue;
+
+				if (x.Actor.Owner != self.Owner || !x.Trait.Info.Produces.Contains(Info.Type))
+					continue;
+
+				Enabled |= IsValidFaction;
+				isActive |= !x.Trait.IsTraitPaused;
+			}
+
+			if (!Enabled)
+				ClearQueue();
+
+			TickInner(self, !isActive);
+		}
+
+		protected override void TickInner(Actor self, bool allProductionPaused)
+		{
+			CancelUnbuildableItems();
+
+			if (allProductionPaused)
+				return;
+
+			var item = Queue.FirstOrDefault(i => !i.Paused);
+			if (item == null)
+				return;
+
+			var parallelBuilds = Queue.FindAll(i => !i.Paused && !i.Done)
+				.GroupBy(i => i.Item)
+				.ToList()
+				.Count - 1;
+
+			if (parallelBuilds > 0 && !developerMode.FastBuild)
+			{
+				penalty -= 100;
+				if (penalty < 0)
+				{
+					penalty = info.ParallelPenalty[Math.Min(parallelBuilds, info.ParallelPenalty.Length - 1)];
+				}
+				else
+					return;
+			}
+
+			var before = item.RemainingTime;
+			item.Tick(playerResources);
+
+			if (item.RemainingTime == before)
+				return;
+
+			// As we have progressed this actor type, we will move all queued items of this actor to the end.
+			foreach (var other in Queue.FindAll(a => a.Item == item.Item))
+			{
+				Queue.Remove(other);
+				Queue.Add(other);
+			}
+		}
+
+		public override IEnumerable<ActorInfo> AllItems()
+		{
+			return Enabled ? base.AllItems() : NoItems;
+		}
+
+		public override IEnumerable<ActorInfo> BuildableItems()
+		{
+			return Enabled ? base.BuildableItems() : NoItems;
+		}
+
+		public override bool IsProducing(ProductionItem item)
+		{
+			return Queue.Contains(item);
+		}
+
+		public override TraitPair<Production> MostLikelyProducer()
+		{
+			var productionActors = self.World.ActorsWithTrait<Production>()
+				.Where(x => x.Actor.Owner == self.Owner
+					&& !x.Trait.IsTraitDisabled && x.Trait.Info.Produces.Contains(Info.Type))
+				.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
+				.ThenByDescending(x => x.Actor.ActorID)
+				.ToList();
+
+			var unpaused = productionActors.FirstOrDefault(a => !a.Trait.IsTraitPaused);
+			return unpaused.Trait != null ? unpaused : productionActors.FirstOrDefault();
```

Comment:
You also need to check if `unpaused` itself is null. Or not use `FirstOrDefault` above.

Reply:
Actually, maybe not. I don't remember if `structs` default to null. So you might ignore this.

Reply:
This was directly taken from https://github.com/OpenRA/OpenRA/blob/23b3c237b7071fd308c4664b0b6c5d719c0f3c74/OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs#L91-L92

Reply:
Ok, thanks for the clarification.